File: apps/customer-mobile/lib/features/chat/data/chat_repository.dart
import 'package:hotel_hall_core/hotel_hall_core.dart';

import 'chat_models.dart';

/// Communication V1 — the backend is authoritative for persistence,
/// read state, and participant scoping (backend/src/modules/chat); this
/// only forwards requests and parses responses, the same shape
/// `NotificationRepository` already follows.
class ChatRepository {
  ChatRepository(this._client);
  final ApiClient _client;

  /// Oldest first (Business Rule 6 — the opposite order from Notification's
  /// own list), matching the backend's own ordering.
  Future<({List<ChatMessage> messages, bool hasNext, String? nextCursor})> list(
    String bookingId, {
    String? cursor,
    int limit = 50,
  }) async {
    final result = await _client.getPaginated(
      '/bookings/$bookingId/messages',
      query: {'limit': '$limit', if (cursor != null) 'cursor': cursor},
    );
    final messages = (result.data as List)
        .map((item) => ChatMessage.fromJson((item as Map).cast<String, dynamic>()))
        .toList();
    return (
      messages: messages,
      hasNext: result.pagination?['hasNext'] as bool? ?? false,
      nextCursor: result.pagination?['nextCursor'] as String?,
    );
  }

  Future<ChatMessage> send(String bookingId, String body) async {
    final data = await _client.post('/bookings/$bookingId/messages', body: {'body': body});
    return ChatMessage.fromJson((data as Map).cast<String, dynamic>());
  }

  Future<void> markConversationRead(String bookingId) => _client.post('/bookings/$bookingId/messages/read-all');

  /// Total unread messages across every Booking the caller participates in
  /// (Business Rule 8) — a second, independent badge from Notification's
  /// own unread count, never merged with it.
  Future<int> unreadCount() async {
    final data = await _client.get('/messages/unread-count');
    return (data as Map)['count'] as int;
  }
}

File: backend/src/modules/chat/chat.repository.js
import { prisma } from '../../shared/prismaClient.js'

/**
 * The only place Prisma Client is called for ChatMessage — no business logic
 * (coding-standards.md §5). `findBookingForParticipant` reaches into the
 * `booking`/`hotel` tables directly (never through `bookings`' own exported
 * functions) the same pragmatic way `notification.repository.js#
 * listPlatformAdministratorIds` already reaches into `user` — a minimal,
 * read-only projection, not a re-implementation of Booking Management's own
 * logic.
 */

export function findBookingForParticipant(bookingId) {
  return prisma.booking.findUnique({
    where: { id: bookingId },
    select: { id: true, customerUserId: true, hotelId: true, hotel: { select: { registeredByUserId: true } } },
  })
}

export function create({ bookingId, senderUserId, body }) {
  return prisma.chatMessage.create({ data: { bookingId, senderUserId, body } })
}

/** Oldest first (Business Rule 6) — the opposite order from Notification's own list. */
export function listForBooking({ bookingId, cursor, take }) {
  return prisma.chatMessage.findMany({
    where: { bookingId },
    orderBy: [{ createdAt: 'asc' }, { id: 'asc' }],
    take,
    ...(cursor ? { skip: 1, cursor: { id: cursor } } : {}),
  })
}

/** Marks every message in this Booking sent by someone other than `readerUserId` read. */
export function markOthersReadInBooking({ bookingId, readerUserId }) {
  return prisma.chatMessage.updateMany({
    where: { bookingId, senderUserId: { not: readerUserId }, readAt: null },
    data: { readAt: new Date() },
  })
}

/**
 * Total unread messages across every Booking `userId` participates in
 * (Business Rule 8) — a second, independent badge source from Notification's
 * own unread count.
 */
export function countUnreadForUser(userId) {
  return prisma.chatMessage.count({
    where: {
      readAt: null,
      senderUserId: { not: userId },
      booking: { OR: [{ customerUserId: userId }, { hotel: { registeredByUserId: userId } }] },
    },
  })
}
